service/domain/blobs/replication/downloader.go

```diff
@@ -0,0 +1,94 @@
+package replication
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/boreq/errors"
+	"github.com/planetary-social/go-ssb/logging"
+	"github.com/planetary-social/go-ssb/service/domain/blobs"
+	"github.com/planetary-social/go-ssb/service/domain/messages"
+	"github.com/planetary-social/go-ssb/service/domain/refs"
+	"github.com/planetary-social/go-ssb/service/domain/transport"
+	"github.com/planetary-social/go-ssb/service/domain/transport/rpc"
+)
+
+type BlobStorage interface {
+	Store(id refs.Blob, r io.Reader) error
+}
+
+type BlobsGetDownloader struct {
+	storage BlobStorage
+	logger  logging.Logger
+}
+
+func NewBlobsGetDownloader(storage BlobStorage, logger logging.Logger) *BlobsGetDownloader {
+	return &BlobsGetDownloader{
+		storage: storage,
+		logger:  logger.New("downloader"),
+	}
+}
+
+func (d *BlobsGetDownloader) OnHasReceived(ctx context.Context, peer transport.Peer, blob refs.Blob, size blobs.Size) {
+	d.logger.WithField("blob", blob.String()).Debug("has")
+
+	if size.Above(blobs.MaxBlobSize()) {
+		d.logger.WithField("size", size).Debug("blob too large")
+		return
+	}
+
+	// todo pre check if we want this blob
```

Comment:
I think that verifying both of these checks could be on the `Manager` side, leaving the `Downloader` purely responsible for downloading. It would save us spawning unnecessary very short living `OnHasReceived` goroutines.

Reply:
I agree but there will have to be some synchronisation going on in the future as we also want to remove blobs. So we have to make sure that the database is in sync with storage somehow. Not sure how to do this yet. Can we leave this for a future PR?